src/l1_fetcher.rs

```diff
@@ -0,0 +1,123 @@
+use ethers::{abi::Contract, prelude::*, providers::Provider};
+use eyre::Result;
+use state_reconstruct::parse_calldata;
+use state_reconstruct::CommitBlockInfoV1;
+use state_reconstruct::BLOCK_STEP;
+use state_reconstruct::GENESIS_BLOCK;
+use state_reconstruct::ZK_SYNC_ADDR;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+
+pub struct L1Fetcher {
+    provider: Provider<Http>,
+    contract: Contract,
+}
+
+impl L1Fetcher {
+    pub fn new(http_url: &str) -> Result<Self> {
+        let provider =
+            Provider::<Http>::try_from(http_url).expect("could not instantiate HTTP Provider");
+
+        let abi_file = std::fs::File::open("./IZkSync.json")?;
+        let contract = Contract::load(abi_file)?;
+
+        Ok(L1Fetcher { provider, contract })
+    }
+
+    pub async fn fetch(
```

Comment:
My thinking with the API here is to separate data fetching & parsing for further processing (including saving it to storage).

This way we can create a command that reads blocks from L1 but only saves them to external file (without utilizing the merkle tree storage). It also allows for reading from L1 and saving to storage, but also just printing as it is right now.